### DISCUSSION TITLE
Project architecture

### DISCUSSION THREAD

**Discussion by gubanovpm** [BGKP-ZI/WLAN-steganography#1] [Ideas]
Here you can write your ideas about code architecture

**Comment by sqwoznyak** [comment]
At the moment we have agreed that there will be two clients and an wireless access point

**Comment by gubanovpm** [comment]
sample project architecture

struct ip_address {
    uint32_t address;
    operator uint32_t(ip_address);
    operator ip_address(uint32_t); 
};
//===========================================================================================
// TODO: using multithreading make sense
//       therefore we must include it into our methods and architecture
//===========================================================================================
//scan current wlan with purpose of getting all wlan users;
class Network final {
public:
    Network(interface &);
    
    void regenerate_networklist();
    void set_new_interface();
    std::string  get_interface();
    void dump_table();
    void create_socket(ip_address);
private:
    interface interface_name;
    std::vector<ip_address> ip_atable;
};
//===========================================================================================
enum class KeyGenerators = {/* Key generator method types */};
class IKeyGenerator {
public:
    excplicit IKeyGenerator(KeyGenerators, seed_type seed) = 0;
    int get_next_key();
private:
    seed_type seed;
};
//===========================================================================================
enum class ChipperMethods = {/* Chipper method types */};
class IChipperMethod {
public:
    IChipperMethod(KeyGenerators);
    std::sstream chipper_data  (std::sstream data);         // ? about returning type
    std::sstream dechipper_data(std::sstream chipper_data); // ? about returning type
    
private:
    IKeyGenerator *key_gerator;
    // mb private and public keys
};
//===========================================================================================
enum class TransferMethods = {/* Transfer method types */};
class ITransferMethod {
public:
    IMethod()
    virtual int send_data(socket_id) = 0;
    virtual int get_data(soscket_id) = 0;
    
private:
};
//===========================================================================================
class client final {
    void data_transform(path_to_file);
    Network network;
    client(TransferMethods, KeyGenerators, ChipperMethods);

private:
    ITransferMethods *transfer_method;
    IKeyGenerator *key_gerator;
    IChipperMethod *chipper_method;
};

**Reply by IvanKhrol** [reply]
the best

**Comment by gubanovpm** [comment]
Semyon:
enum class KeyGenerators = {/* Key generator method types  as a exmaple BBS - Blum Blum Shub*/};
class IKeyGenerator

**Comment by gubanovpm** [comment]
Project structure:

include 
-->Graphics
  ---->Graphics.hh
  ----> //other header files
--> Crypto
  ----> Crypto.hh
  ----> //other header files
-->Network
  ----> Network.hh
  ----> //other header files
-->//other include dirs

sources
-->Graphics
  ---->Graphics.cc
  ----> //other files
  ----> CMakeLists.txt
--> Crypto
  ----> Crypto.cc
  ----> //other files
  ----> CMakeLists.txt
-->Network
  ----> Network.cc
  ----> //other files
  ----> CMakeLists.txt
-->//other source dirs

samples
--> test_1
  ----> test_1.cc
  ----> CMakeLists.txt
--> //other samples directories

libs
--> //compiled libs

build
-->samples
  ----> test_1
     ------> test_1.out
  ----> ...
  ----> release
    ------> wlan_steg.exe

CMakeLists.txt

**Comment by gubanovpm** [comment]
Ivan: 
Graphics for application
